Add Relay option to link parameters into runtime Modules

areusch opened the pull request:
Implements RFC: https://discuss.tvm.apache.org/t/rfc-linked-parameters-for-cpu-targets/8452

This PR implements changes to support generating Modules that contain parameters pre-linked. For CPU contexts, these modules don't require any copying before launching inference via GraphRuntime. Both C++ and C runtimes are supported.

areusch commented:
@tqchen @u99127 @tom-gall @manupa-arm @leandron

tqchen commented:
cc @jroesch @junrushao1994

manupa-arm commented on src/relay/backend/graph_runtime_codegen.cc:
```diff
@@ -56,7 +56,7 @@ struct LoweredOutput {
   Map<String, IRModule> lowered_funcs;
   Array<tvm::runtime::Module> external_mods;
-  std::unordered_map<std::string, tvm::runtime::NDArray> params;
+  std::unordered_map<std::string, std::pair<int, const tvm::runtime::NDArray>> params;
```
This change will affect UpdateConstants() in line 379. Either we need to modify the UpdateConstants function or have something similiar to param_storage_ids in LoweredOutput separately

manupa-arm commented on python/tvm/contrib/binutils.py:
```diff
@@ -16,61 +16,10 @@
-
-# TODO shouldn't need so many `ALIGN` directives
-RELOCATION_LD_SCRIPT_TEMPLATE = """
-/* linker symbol for use in UTVMInit */
```
Is this change in the scope of the PR/RFC ? if not suggest we break it out to another PR.

areusch replied:
broken out

manupa-arm commented on include/tvm/runtime/crt/platform.h:
```diff
@@ -39,6 +41,21 @@ extern "C" {
  */
 void __attribute__((noreturn)) TVMPlatformAbort(tvm_crt_error_t code);
 
+/*! \brief Called by the microTVM RPC server to implement TVMLogf.
```
Is this change relevant to the PR/RFC? If not suggest break it out to a different PR.

areusch replied:
broken out and submitted separately

manupa-arm commented on include/tvm/runtime/crt/graph_runtime.h:
```diff
@@ -61,14 +61,20 @@ typedef struct TVMGraphRuntime TVMGraphRuntime;
+ * \brief get number of input tensors allocated.
+ * \return integer number of tensors available to use.
+ */
+int TVMGraphRuntime_GetNumInputs();
```
Maybe its just me, but I cant see how this is relevant to linked params ?

areusch replied:
this one is needed for the CRT test. we didn't yet support the case of executing the graph runtime within the C runtime over RPC. test_linked_params uses this to verify that the CRT graph runtime can lookup the linked params. it's a bit tangential yes, but it is basically rounding out the graph runtime impl and it's not a very large addition; i'd prefer to keep it with this change to show the motivation.

manupa-arm replied:
fine by me, then.

manupa-arm commented on include/tvm/runtime/crt/graph_runtime.h:
```diff
@@ -77,6 +83,12 @@ int TVMGraphRuntime_GetInputIndex(TVMGraphRuntime* runtime, const char* name);
+ * \brief get number of output tensors allocated.
+ * \return integer number of output tensors allocated.
+ */
+int TVMGraphRuntime_GetNumOutputs();
```
Maybe its just me, but I cant see how this is relevant to linked params ?

areusch replied:
same story as before

manupa-arm reviewed (changes requested):
Hi Andrew,

Thanks for the change and I think its quite useful for linked params to be situated in RO data sections.

I had to go through it looks you are testing uTVM with linked params which is good!. However, I feel there are certain error handling enhancement are there not exactly relevant to the scope of the PR (I might be wrong or failing to see the connections -- if so feel free to say so :) ). While I personally dont disagree with such changes, but would like to request to break them out of the PR to help the reviewers if its not too much trouble from your end.

tqchen commented:
cc @kparzysz-quic

tqchen commented on src/target/llvm/codegen_llvm.cc:
```diff
@@ -184,6 +188,104 @@ void CodeGenLLVM::AddFunctionInternal(const PrimFunc& f, bool ret_void) {
 }
 
+void CodeGenLLVM::LinkParameters(const Map<String, LinkedParam> params) {
+  // It would be nice to de-dupe these declarations frm src/tir/transforms/make_packed_api.cc,
```
It is possible to list more into make_packed_api later, if we have intrinsics like

```
value = tir.lookup_linked_param_in_attr(id)

```

The next steps should be supported once we have return value in PackedFunc. cc @ZihengJiang

ZihengJiang replied:
I see

tqchen commented on src/target/llvm/codegen_params.cc:
```diff
@@ -0,0 +1,151 @@
+  for (auto shape_elem : shape) {
+    num_elements *= shape_elem;
+  }
+
```
I don't think we need to convert things to DLPack. arr.operator-> already gives you a DLTensor*

areusch replied:
done

tqchen commented on src/target/llvm/codegen_params.cc:
```diff
@@ -0,0 +1,151 @@
+                                << arr_type.lanes();
+
+  auto shape = arr.Shape();
+  int num_elements = 1;
```
Need to CHECK arr IsContiguous() and is on CPU

areusch replied:
done

tqchen commented on src/target/llvm/codegen_params.cc:
```diff
@@ -0,0 +1,151 @@
+
+llvm::ConstantArray* NDArrayToLLVMArray(llvm::LLVMContext* ctx, ::tvm::runtime::NDArray arr) {
+  llvm::Type* element_type = nullptr;
+
```
I wonder whether or not should we instead look into LLVM's target machine to lookup the endian parameter, and do byte swap ourselves.

This would allow us to generalizes to data types that goes beyond the types that can recognized by LLVM, as long as we know the target machine layout

See https://github.com/apache/incubator-tvm/blob/main/src/runtime/rpc/rpc_endpoint.cc#L429

areusch replied:
the thing is, because it will be used directly by instructions (I.e. load half-word), I prefer to let LLVM deal with alignment and whatever else is necessary to make that work. my understanding was we tried an experiment compiling large constants and the bulk of the time was spent in parsing, so this shouldn't actually load LLVM too much? also, we just do this a few times for the whole model

tqchen replied:
How about we take the middle ground approach, e.g. using the same data type for the same bits. E.G. use uint32 for both float32, int32. This will allow us to effectively support new data types without having llvm to be aware of it

areusch replied:
won't we need to expand the switch-case statement to support new datatypes anyhow? for some reason, LLVM has a different constant impl for FP and int constants, so I guess i sort of like the approach taken now in the absence of any example new datatypes. i also don't think it's inflexible to supporting those new datatypes. note that LLVM does treat signed and unsigned constants as the same; just the codepath to instantiate them is slightly different.

tqchen replied:
We still need to support cases like bfloat16, which is not natively supported by LLVM. It would still be great to support via uint in these cases

areusch replied:
added support for float16 and bfloat16 as discussed.

tqchen commented:
cc  @ZihengJiang would be great if you can also take a look as the change relates to tir and PackedFunc codegen

tqchen commented:
I did a pass on the C++ side(didn't get a chance to look at c module) the main thing to pay attentions are:

- NDArray's deleter handling, need to delete the container itself as well(besides the resources)
- Get DLTensor repr does not need ToDLPack, instead, just use NDArray.operator->

Things to discuss:
- the type switch that generate typed LLVM array can be quite long, and requires creation of a lot of LLVM objects. If we can use LLVM's target mechanism to know the endian of the target machine, perhaps it is better to do byteswaping by ourselves. Alternatively, we can simply only switch on bits, but not types.

E.g. use uint32 for both float32, int32, and uint32 array.

areusch commented on src/target/llvm/codegen_params.cc:
```diff
@@ -0,0 +1,151 @@
+      element_type = llvm::Type::getIntNTy(*ctx, arr_type.bits());
+
+      if (arr_type.bits() == 8) {
+        int8_t* data_buf = static_cast<int8_t*>(tensor->dl_tensor.data);
```
not in this one. I had originally kept them handwritten separately while developing in case more things needed to be changed, and for speed. I agree, template is more maintanable here. done.

areusch commented on src/target/llvm/codegen_params.cc:
```diff
@@ -0,0 +1,151 @@
+      break;
+
+    case runtime::DataType::TypeCode::kUInt:
+      CHECK(arr_type.bits() == 8 || arr_type.bits() == 16 || arr_type.bits() == 32 ||
```
done

areusch commented on src/target/llvm/codegen_params.cc:
```diff
@@ -0,0 +1,151 @@
+
+    case runtime::DataType::TypeCode::kFloat:
+      if (arr_type.bits() == 32) {
+        element_type = llvm::Type::getFloatTy(*ctx);
```
done

areusch commented on include/tvm/tir/function.h:
```diff
@@ -150,6 +151,31 @@ class PrimFunc : public BaseFunc {
   TVM_DEFINE_OBJECT_REF_COW_METHOD(PrimFuncNode);
 };
 
+class LinkedParamNode : public Object {
```
done

areusch commented on include/tvm/tir/function.h:
```diff
@@ -150,6 +151,31 @@ class PrimFunc : public BaseFunc {
+
+class LinkedParam : public ObjectRef {
+ public:
+  LinkedParam(int64_t id, ::tvm::runtime::NDArray param);
```
done

areusch commented on include/tvm/tir/function.h:
```diff
@@ -150,6 +151,31 @@ class PrimFunc : public BaseFunc {
+  TVM_DECLARE_FINAL_OBJECT_INFO(LinkedParamNode, Object);
+};
+
+class LinkedParam : public ObjectRef {
```
done

areusch commented on src/relay/backend/graph_runtime_codegen.cc:
```diff
@@ -582,7 +596,13 @@ class GraphRuntimeCodegenModule : public runtime::ModuleNode {
+    } else if (name == "get_param_id") {
+      return PackedFunc([sptr_to_self, this](TVMArgs args, TVMRetValue* rv) {
+        String key = args[0];
+        ICHECK_GT(this->output_.params.count(key), 0);
```
done

areusch commented on src/runtime/graph/graph_runtime.cc:
```diff
@@ -244,6 +249,39 @@ void GraphRuntime::ShareParams(const GraphRuntime& other, dmlc::Stream* strm) {
+  TVMContext ctx = args[3];
+  // Get pre-linked parameter lookup function, if it was generated. When pf == nullptr, no linked
+  // params are present.
+  tvm::runtime::PackedFunc pf =
```
done

areusch commented on src/runtime/graph/graph_runtime.cc:
```diff
@@ -64,14 +64,19 @@ void GraphRuntime::Run() {
   ctxs_ = ctxs;
+  lookup_linked_param_ = lookup_linked_param_func;
+  if (lookup_linked_param_ == nullptr) {
+    lookup_linked_param_ = PackedFunc(&GraphRuntime::DefaultLookupLinkedParam);
```
done

areusch commented on src/runtime/graph/graph_runtime.cc:
```diff
@@ -64,14 +64,19 @@ void GraphRuntime::Run() {
  */
 void GraphRuntime::Init(const std::string& graph_json, tvm::runtime::Module module,
-                        const std::vector<TVMContext>& ctxs) {
+                        const std::vector<TVMContext>& ctxs, PackedFunc lookup_linked_param_func) {
```
the return type is nullptr or int, so keeping untyped.

areusch commented on src/runtime/graph/graph_runtime.cc:
```diff
@@ -244,6 +249,39 @@ void GraphRuntime::ShareParams(const GraphRuntime& other, dmlc::Stream* strm) {
   this->SetupOpExecs();
 }
 
+void GraphRuntime::PreAllocatedDLTensorDeleter(DLManagedTensor* tensor) {
```
ah you are right! changed.

areusch commented on src/runtime/graph/graph_runtime.cc:
```diff
@@ -244,6 +249,39 @@ void GraphRuntime::ShareParams(const GraphRuntime& other, dmlc::Stream* strm) {
+  std::vector<int64_t> shape_vec{template_tensor->shape,
+                                 template_tensor->shape + template_tensor->ndim};
+
+  std::unique_ptr<NDArray::Container> container{new NDArray::Container(
```
done

areusch commented on src/runtime/rpc/rpc_module.cc:
```diff
@@ -36,6 +37,44 @@
+ */
+NDArray NDArrayFromRemoteOpaqueHandle(std::shared_ptr<RPCSession> sess, void* handle,
+                                      DLTensor* template_tensor, TVMContext ctx,
+                                      ADTObj::FDeleter deleter, void* deleter_ctx) {
```
fair enough, done.

areusch replied:
i've changed this so if the handle is nullptr, it won't delete. I thiink that will probably work, let me know your thoughts.

areusch commented on src/target/llvm/codegen_llvm.cc:
```diff
@@ -184,6 +188,104 @@ void CodeGenLLVM::AddFunctionInternal(const PrimFunc& f, bool ret_void) {
+  param_types.push_back(t_void_->getPointerTo(GetGlobalAddressSpace()));
+
+  // TODO(tvm-team):
+  // Update the function type to respect the ret_type field of f.
```
ah, copied that from somewhere, deleted.

areusch commented on src/target/llvm/codegen_llvm.cc:
```diff
@@ -184,6 +188,104 @@ void CodeGenLLVM::AddFunctionInternal(const PrimFunc& f, bool ret_void) {
+        llvm::ArrayType::get(t_int_, 1)->getPointerTo());
+
+    builder_->CreateStore(llvm::ConstantInt::get(t_int_, kTVMNullptr),
+                          builder_->CreateGEP(ret_types_array, zero_array_index_list));
```
done

areusch commented on src/target/llvm/codegen_llvm.cc:
```diff
@@ -184,6 +188,104 @@ void CodeGenLLVM::AddFunctionInternal(const PrimFunc& f, bool ret_void) {
+    builder_->CreateRet(ConstInt32(kTvmErrorNoError));
+  }
+
+  llvm::raw_os_ostream os{std::cout};
```
done

areusch commented on src/target/llvm/codegen_llvm.cc:
```diff
@@ -184,6 +188,104 @@ void CodeGenLLVM::AddFunctionInternal(const PrimFunc& f, bool ret_void) {
+                           builder_->CreateInBoundsGEP(args_array, zero_index_list)),
+      t_int64_);
+
+  llvm::BasicBlock* default_block = llvm::BasicBlock::Create(*ctx_, "default_block", function);
```
default like "default:" in a switch case

areusch commented on src/target/llvm/codegen_llvm.cc:
```diff
@@ -184,6 +188,104 @@ void CodeGenLLVM::AddFunctionInternal(const PrimFunc& f, bool ret_void) {
+
+  builder_->SetInsertPoint(default_block);
+  {
+    auto ret_types_array = builder_->CreateBitCast(
```
I don't follow?

areusch replied:
nvm, done

areusch commented on src/target/llvm/codegen_llvm.cc:
```diff
@@ -184,6 +188,104 @@ void CodeGenLLVM::AddFunctionInternal(const PrimFunc& f, bool ret_void) {
+        llvm::cast<llvm::ConstantInt>(llvm::ConstantInt::get(t_int64_, kv.second->id)), case_block);
+    builder_->SetInsertPoint(case_block);
+    auto retval_array = builder_->CreateBitCast(
+#if TVM_LLVM_VERSION >= 50
```
done

areusch commented:
I found a fairly serious bug, now fixed, on more complex networks. we weren't quite pulling the storage_id for each param; it just typically happens to match the parameter node id in the graph. reworked my test a bit to expose this bug and fixed.

@manupa-arm @tqchen @ZihengJiang please take another look when you have a minute

kparzysz-quic commented on src/target/source/codegen_params.cc:
```diff
@@ -0,0 +1,297 @@
+    case runtime::DataType::TypeCode::kFloat: {
+      std::stringstream ss;
+      ss.setf(std::ios::hex | std::ios::showbase | std::ios::fixed | std::ios::scientific,
+              std::ios::basefield | std::ios::showbase | std::ios::floatfield);
```
`std::showbase` is there twice.

areusch replied:
right, the second one is a mask so I think it's needed twice?

kparzysz-quic commented on src/target/source/codegen_params.cc:
```diff
@@ -0,0 +1,297 @@
+  CHECK_EQ(arr_type.lanes(), 1) << "CodegenParams: only support generating 1-lane parameters; saw "
+                                << arr_type.lanes();
+
+  int one_element_size_bytes = (arr_type.bits() / 4) + (2 /* "0x" */) + (2 /* ", " */);
```
In hexadecimal floating point, the exponent is still in decimal, only the significant is in hex.  This calculation (the `bits/4` part) isn't entirely correct for fp.

kparzysz-quic commented on src/target/llvm/codegen_llvm.cc:
```diff
@@ -184,6 +188,90 @@ void CodeGenLLVM::AddFunctionInternal(const PrimFunc& f, bool ret_void) {
+      llvm::ArrayType::get(t_void_->getPointerTo(GetGlobalAddressSpace()), 1));
+  llvm::Value* sid = builder_->CreateBitCast(
+      builder_->CreateLoad(t_void_->getPointerTo(GetGlobalAddressSpace()),
+                           builder_->CreateInBoundsGEP(args_array, zero_index_list)),
```
You don't need this GEP.

areusch replied:
are you sure including on LLVM 4? I guess i'm wondering if it's hurting anything...I think it is technically correct to be there, even if it is a no-op?

areusch commented on src/target/llvm/codegen_params.cc:
```diff
@@ -0,0 +1,172 @@
+template <typename T, typename = std::enable_if<std::is_pod<T>::value>>
+void BuildLLVMVector(llvm::Type* element_type, void* tensor_data, size_t num_elements,
+                     std::vector<llvm::Constant*>* elements) {
+  for (size_t i = 0; i < num_elements; i++) {
```
done

areusch commented on src/target/source/codegen_params.cc:
```diff
@@ -0,0 +1,297 @@
+
+  int one_element_size_bytes = (arr_type.bits() / 4) + (2 /* "0x" */) + (2 /* ", " */);
+  if (arr_type.code() == runtime::DataType::TypeCode::kInt) {
+    one_element_size_bytes += 1;  // sign bit
```
done

areusch commented on src/target/source/codegen_params.cc:
```diff
@@ -0,0 +1,297 @@
+    one_element_size_bytes += 1;  // sign bit
+    if (arr_type.bits() > 32) {
+      one_element_size_bytes += 2;  // "UL"
+    }
```
changed to "LL" here

areusch commented on src/target/source/codegen_params.cc:
```diff
@@ -0,0 +1,297 @@
+          << "CodegenParams: only support generating 8-, 16-, 32-, or 64-bit integer params; saw "
+          << arr_type.bits() << "-bit array";
+
+      if (arr_type.bits() == 8) {
```
ah great. refactored into templatized functions.

areusch commented:
@kparzysz-quic addressed those comments.
@manupa-arm @tqchen @ZihengJiang please take another look when you have a minute and explicit approve if you're ok with this change.

tqchen commented on src/target/source/codegen_params.cc:
```diff
@@ -0,0 +1,214 @@
+      os.fill(' ');
+      os.setf(std::ios::left, std::ios::adjustfield);
+      if (arr_type.bits() == 32) {
+        PrintArray<float>(tensor->dl_tensor.data, num_elements, one_element_size_bytes,
```
I still think it is easier to use uint to print the case of float in most of our cases. It also helps us to get around the problem of printing exp and mantisa

areusch replied:
I mean, the printing problem is just an alignment thing. it doesn't affect the functionality. I don't see why adding an implicit cast to generated code that the compiler will never even know about is easier? because remember, we are not going to add the cast into the generated function that is accessing these floats--it is just going to get a float* and then by generating uint32_t here, the float* will magically access uint32_t data. I agree with you that it doesn't seeeem like anything would break too badly given they are the same size, but it's kind of a long limb to go out on just to support bfloat16. we would also need to do byte ordering.

tqchen replied:
To put it in another way. We will need support for bfloat16 and fp16 now given that they are already part of the type system. Addtiionally, there might also be need to support customized data types. 

The main reason to go for the uint repr is its bit accurate precision(we don't need to worry about printing out INF, nan, or loss of accuracy due to printing and parsing of floating pt -- although partly addressed by hex printing, still somewhat complicated) and simplicity in implementation.

Additionally, for big endian machines and small endian machines we are already using the bits as the only information in our runtime https://github.com/apache/tvm/blob/main/include/tvm/runtime/ndarray.h#L429, which means our runtime won't simply work beyond such cases. Right now we also did not see an example that goes beyond these two cases.

If we do think it is important to keep float and double printing as their own type. I think we could do that, but need to make sure the rest of the codepath uses uint so that it works for bfloat16, fp16 and other potential extension data types now.

areusch replied:
okay, as discussed i've added support for float16 and bfloat16. left the rest of the float printing as it was before.

tqchen commented:
I still think we can print all values by their bits and print via uint. Unless there is a very strange endian being used. It would also helps us to support types that are not supported by LLVM natively, like bfloat16

areusch commented:
@tqchen it seems like the main benefit is supporting bfloat16...i'd like to argue that's out of scope for this PR. further, we don't really have a big-endian board handy to test against anyway. I don't think that change would take a long time to implement, but we don't have a diverse enough test fleet now to have good confidence in it.

FrozenGene commented on pyproject.toml:
```diff
@@ -46,3 +46,95 @@ exclude = '''
+importer-onnx = ["onnx", "onnxruntime", "torch", "torchvision", "future"]
+importer-pytorch = ["torch", "torchvision", "future"]
+importer-tensorflow = ["tensorflow", "tensorflow-estimator"]
+importer-tflite = ["tlfite", "tensorflow", "tensorflow-estimator"]
```
tlfite->tflite

areusch commented on src/relay/backend/build_module.cc:
```diff
@@ -443,16 +456,35 @@ class RelayBuildModule : public runtime::ModuleNode {
+    const runtime::PackedFunc* pf = runtime::Registry::Get("codegen.LLVMModuleCreate");
+    if (!target_host.defined()) target_host = (pf != nullptr) ? Target("llvm") : Target("stackvm");
+
+    if (target_host->GetAttr<Bool>("link-params").value_or(Bool(false))) {
```
done

areusch commented on src/target/source/codegen_params.cc:
```diff
@@ -0,0 +1,297 @@
+    }
+  } else if (arr_type.code() == runtime::DataType::TypeCode::kUInt) {
+    if (arr_type.bits() > 32) {
+      one_element_size_bytes += 1;  // "L"
```
changed to "ULL" here

manupa-arm commented on src/target/source/codegen_params.cc:
```diff
@@ -0,0 +1,218 @@
+  }
+}
+
+void NDArrayDataToC(::tvm::runtime::NDArray arr, int indent_chars, std::ostream& os) {
```
Do we need to have this ASCII encoded as such ? Im thinking of why this cant be a raw binary string and get rid of "0x", commas and brackets. IIUC, the DLTensor just needs a void* and would be casted to appropriate data type in the function in which it consumes the constant data.

If the reasoning is for debugging then probably we should have a debug flag and do this if required and default could be a raw binary string (if thats possible), I think it will save compilation time.

manupa-arm commented on src/target/source/codegen_params.cc:
```diff
@@ -0,0 +1,218 @@
+  }
+}
+
+void NDArrayDataToC(::tvm::runtime::NDArray arr, int indent_chars, std::ostream& os) {
```
It would be great if a brief could be added to show how a simple NDArray would get converted and appear in the source.

areusch replied:
documented this function in codegen_params.h and added an example there.

manupa-arm reviewed (commented):
I did a second pass. Thanks for breaking out some of it as a separate PR.
Some comments and some of it might end up being clarifications.

manupa-arm commented on src/target/source/codegen_params.cc:
```diff
@@ -0,0 +1,218 @@
+  }
+}
+
+void NDArrayDataToC(::tvm::runtime::NDArray arr, int indent_chars, std::ostream& os) {
```
[Clarification] Can we use a raw binary string (e.g. "\xeb\x2a) to represent the content here ? Do we need to show the element wise break down? I would reckon DLTensors would anyway take it as a void* right ?

areusch replied:
we could. the main thing this drags in is whether we should be determining things like the byte order and alignment of the target machine, or leave that to the compiler (LLVM or in this case, external C). i've been so far arguing not to do any of this--I don't really see a tangible benefit and it adds considerable complexity at this level that's hard to test. also, it makes it quite easy to inspect the generated parameters, which will often differ from the parameters originally supplied to relay.build().

areusch commented on src/target/source/codegen_c_host.cc:
```diff
@@ -57,6 +60,48 @@ void CodeGenCHost::AddFunction(const PrimFunc& f) {
+           << "        ((uint64_t*)out_ret_value)[0] = (uint64_t) (uintptr_t) "
+           << ::tvm::runtime::symbol::tvm_param_prefix << kv.first << ";\n"
+           << "        out_ret_tcode[0] = " << kTVMOpaqueHandle << ";\n"
+           << "        return 0;\n";
```
it would be nice to return a specific error code here (this was my original implementation), but unfortunately it's not easy to catch on the other side. we need to do a bit more legwork to be able to catch specific function return values in c++/python. as it stands, returning nullptr from this function is specific enough, and has the benefit that an exception flow isn't triggered by default for non-parameters.

areusch commented on src/target/source/codegen_c_host.cc:
```diff
@@ -57,6 +60,48 @@ void CodeGenCHost::AddFunction(const PrimFunc& f) {
+                << "}  // extern \"C\"\n"
+                << "#endif\n";
+    stream << "    case " << kv.second->id << ":\n"
+           << "        ((uint64_t*)out_ret_value)[0] = (uint64_t) (uintptr_t) "
```
it's just copying the pointer here. I believe that should be ok?

manupa-arm replied:
Oh ok. So kv,first is the array's name, right ? Any reason why is it casted to uint64_t ? Im thinking of a non 64 bit platform usecase.

areusch replied:
correct. ordinarily we would cast to void*, but since void* may be less than 64 bits, we cast to uint64_t which is the largest type stored in TVMValue.

manupa-arm replied:
I still think void* should be enough here. void* would be less than 64 bits if the machine compiled for is less than 64bits. right? otherwise this will make it store the pointer in 64bit space where the machine could have a address space defined less than 64 bits, unless Im missing something here. In this specific case, the array defined for the constant will have an address depending the target machine it would be compiled for, hence my confusion.

Do we need to perform pointer arithmetic on this ? -- Im asking this because uintptr_t cast is used. If so, using uintptr_t should be sufficient instead of void*.

areusch replied:
TVMValue is always 64 bits wide, though, because it's a union including int64_t. so when addressing it as an array, need to cast appropriately. i'm not sure if the uintptr_t cast is excessive--quite possibly it is.

manupa-arm replied:
Alright then. I will take your word for it. :). See if you can drop uintptr_t if that is not required given that it is going to be written to 64bit wide space location anyway.

tqchen commented on src/target/source/codegen_params.cc:
```diff
@@ -0,0 +1,218 @@
+    num_elements *= shape_elem;
+  }
+
+  std::unique_ptr<DLManagedTensor, DLManagedTensorDeleter> tensor(arr.ToDLPack());
```
We don't need ToDLPack, and can directly get DLTensor* from the arr.operator->

areusch replied:
fixed

tqchen reviewed (changes requested):
The NDArray deletion part LGTM, see more discussions on the printing repr and whether or not w should use uint

tqchen commented on src/runtime/graph/graph_runtime.cc:
```diff
@@ -244,6 +250,43 @@ void GraphRuntime::ShareParams(const GraphRuntime& other, dmlc::Stream* strm) {
+void GraphRuntime::LinkedNDArrayDeleter(Object* container) {
+  // container is the NDArray::Container which needs to get deleted.
+  // The data member points to global const memory, so it does not need deleting.
+  delete reinterpret_cast<NDArray::Container*>(container);
```
Should be static cast instead, since NDArray::Container* is a subclass of Object*, this can cause bug when Container and Object have a different offset

areusch replied:
ah, thanks. done.

manupa-arm reviewed (approved):
Mostly LGTM bar others' pending comments.

tqchen commented:
LGTM now, @FrozenGene please take another look

tqchen commented:
Thanks @areusch @ZihengJiang @FrozenGene  @kparzysz-quic @manupa-arm
